content/blog/2019-01-31-rdhs.md

```diff
@@ -0,0 +1,787 @@
+response as the value. Initially I thought I would have to keep saving the response 
+with explicit names (e.g. the URL), but `storr` handles all this for you, and also
+then helps get around having too long file names if your API request is very long for example.
+
+To respond to changes in the API, my solution was perhaps not the neatest, but I 
+simply kept a record of the date you last made an API request and compared it to 
+the API's [data updates endpoint](https://api.dhsprogram.com/#/api-dataupdates.cfm). 
+If I could see any recent changes, I then could clear all the API requests cached. 
+This would made a lot simpler using the `namespaces` options in `storr`, which meant
+that I was able to keep all API cached data in one place, which could then be 
+easily deleted on mass. 
+
+#### 3. Tests, Travis & Authentication
+
+The last thing caused me the most amount of headaches. How do I write tests that
+require authentication and can use `travis` for continuous integration. Initially,
+I made a dummy account with the DHS website for this, but realised that sharing
+the credentials of an account with access to just dummy data sets would not enable
+me to test the weird edge cases that started popping up related to certain data 
+sets. The first solution that I used for a few months was to set up environment 
+variables within `travis` itself, which could then be used to create a valid
+set of credentials. 
+
+![](/img/blog-images/2019-01-31-rdhs/travis_env_vars.png)
+
+This worked, however, it meant that I would have to write a lot of the `rdhs`
+functionality to use environment variables that were the user's email and password, 
+which felt wrong and quite clunky. All I wanted was to pass to Travis was a valid
+set of login credentials that would then be used within the tests, much in the same
+way that a user would. To do this I had to learn a bit more about what the `.travis.yml`
+document could actually be used for, because to begin with I had only been using it
+to specify the software language.
+
+Again, Rich pointed me to using `sodium` to create an encrypted version of a valid
+login credentials:
+
+
+```r
+# read in a key from a local file
+key <- sodium::hash(charToRaw(readLines("scripts/key.txt")))
+
+# create a tat with all the necessary login credentials
+zip("rdhs.json.tar",files=c("rdhs.json", "tests/testthat/rdhs.json"))
+
+# read this tar in as binary data 
+dat <- readBin("rdhs.json.tar",raw(),file.size("rdhs.json.tar"))
+
+# encrypt the data using sodium and our key before saving it
+enc <- sodium::data_encrypt(msg = dat,key = key)
+saveRDS(enc,"rdhs.json.tar.enc")
+```
+
+This could then be included in the GitHub repository, and I could then
+set up the key as a Travis environment variable to decrypt it. This decryption 
+step could then be written within my `.travis.yml` file, and would mean that all
+my tests had access to my login credentials in a secure way. 
+
+### Acknowledgements and Final Thoughts
```

Comment:
This section has a really nice genuine tone

Reply:
I think it would be better to have this section last. The sentiment is great, and "Options to Contribute" leaves the post with an abrupt end